---
title: "DNA_Sequence_Statistics_2"
author: "David_Halvorsen"
date: "August 9, 2018"
output: html_document
---

```{r}
# What's the logbase10 of 100?
x <- 100
log10(x)
# What's the mean of myvector?
myvector <- c(30,16,303,99,11,111)
mean(myvector)
```

```{r}
# extracting a single element from myvector
myvector[3]
```

```{r}
# how to create a sequence of numbers
seq(1, 100, by = 1)
# changing the 'by' argument from 1 -> 2
seq(1, 100, by = 2)
```

```{r}
# using a for loop to print out the square of each number from 1 to 10
for (i in 1:10) { print (i*i) }
```

```{r}
# using a for loop to print the square of each item in a vector
avector <- c(2, 9, 100, 133)
for (i in avector) { print (i*i) }
```

```{r}
# use a for loop to print out the square of every *SECOND* number between 1 and 10?
for (i in seq(1, 10, by = 2)) { print (i*i) }
```

```{r}
# plotting myvector2 and a function of myvector1
myvector1 <- c(10, 15, 22, 35, 43)
myvector2 <- c(3, 3.2, 3.9, 4.1, 5.2)
plot(myvector1, myvector2, xlab="myvector1", ylab="myvector2")
```

```{r}
# adding a type=b argument will draw a line between each point
plot(myvector1, myvector2, xlab="myvector1", ylab="myvector2", type="b")
```

```{r}
# creating a function that returns 20 plus the square of the input
myfunction <- function(x) { return(20 + (x*x)) }
# testing out that function with two inputs
myfunction(10)
myfunction(25)
# printing the entirety of the function
myfunction
# finds the log to the base 10 of variable x
x <- 100
log10(x)

```

```{r}
# reading in a fasta file and selecting JUST the nucleotides
library("seqinr")
dengue <- read.fasta(file="den1.fasta")
dengueseq <- dengue[[1]]
# extracting a subsequence
dengueseq[452:535]
```

```{r}
# what's the GC content of the dengue?
GC(dengueseq)
```

```{r}
# how do I get a rolling window of GC content?
# length is 10735
length(dengueseq)
GC(dengueseq[1:2000])
GC(dengueseq[2001:4000])
GC(dengueseq[4001:6000])
GC(dengueseq[6001:8000])
GC(dengueseq[8001:10000])
GC(dengueseq[10001:10735])
```

```{r}
# it's possible to automate the GC rolling window with a function!
# create the different starting points for the function 
starts <- seq(1, length(dengueseq)-2000, by = 2000)
# I was trying to figure out if this was a vector, BUT it kept telling me the number type
typeof(starts)
# it responds the this bracket command, so it must be a vector, right?
starts[1]
# is.vector checks if something is a vector! It is a vector :)
is.vector(starts)

# determining how long the function will run for
n <- length(starts)
# for 1 - 5
for (i in 1:n) {
  # chunk will be a block of dengueseq wherever the i and i + 1999 determine
  chunk <- dengueseq[starts[i]:(starts[i]+1999)]
  # getting GC info for current chunk
  chunkGC <- GC(chunk)
  print(chunkGC)
}

```

```{r}
# this is the x axis vector ... I didn't realize that when I was doing the first part.
starts <- seq(1, length(dengueseq)-2000, by = 2000)
n <- length(starts)
# this is 1st diff from above: creating an empty vector to store the GC values
chunkGCs <- numeric(n)
for (i in 1:n) {
  chunk <- dengueseq[starts[i]:(starts[i]+1999)]
  chunkGC <- GC(chunk)
  print(chunkGC)
  chunkGCs[i] <- chunkGC
}
plot(starts,chunkGCs,type="b",xlab="Nucleotide start position",ylab="GC content")
```

```{r}
# creating a function that allows the user to select the input sequence and the sliding window size
# the only real difference between the previous function and the changeable one is this next line
# we accept a windowsize and an inputsequence
slidingwindowplot <- function(windowsize, inputseq)
{
  # note that the '2000' from the previous function is replaced by windowsize here
  starts <- seq(1, length(inputseq)-windowsize, by = windowsize)
  n <- length(starts)
  chunkGCs <- numeric(n)
  for (i in 1:n){
    # window size again determining the range (for GC to run on)
    chunk <- inputseq[starts[i]:(starts[i]+windowsize-1)]
    chunkGC <- GC(chunk)
    print(chunkGC)
    chunkGCs[i] <- chunkGC
  }
  plot(starts,chunkGCs,type="b",xlab="Nucleotide start position",ylab="GC content")
}
```

```{r}
# calling that function
slidingwindowplot(3000, dengueseq)
```

```{r}
# calling that function again
slidingwindowplot(300, dengueseq)
```

```{r}
# get all the two-character long 'words' from the dengue sequence
count(dengueseq, 2)
```

```{r}
# determining ro for GC
# first get counts for individual nucleotides
count(dengueseq, 1)
# frequency of G
2770/(3426+2240+2770+2299)
# frequency of C
2240/(3426+2240+2770+2299)
# counts for dinucleotides
count(dengueseq,2)
# frequency of GC dinucleotide
500/(1108 + 720 + 890 + 708 + 901 + 523 + 261 + 555 + 976 + 500 + 787 + 507 + 440 + 497 + 832 + 529 )
# fGC / fG * fC
0.04658096/(0.2580345*0.2086633)
```

#Q1. Draw a sliding window plot of GC content in the DEN-1 Dengue virus genome, using a window size of 200 nucleotides. Do you see any regions of unusual DNA content in the genome (eg. a high peak or low trough)?
```{r}
# using the function I created above.
# very low GC near the beginning, 2500, 3900, 8000, and 1100 ... super high around 13000
slidingwindowplot(200, dengueseq)
slidingwindowplot(2000, dengueseq)
dengueseq
```

#Q2. Draw a sliding window plot of GC content in the genome sequence for the bacterium Mycobacterium leprae strain TN (accession NC_002677) using a window size of 20000 nucleotides. Do you see any regions of unusual DNA content in the genome (eg. a high peak or low trough)?
```{r}
require('seqinr')
# reading in the sequence (got it from ncbi)
read_leprae <- read.fasta(file="/media/david/Linux/R_for_Bioinformatics/DNA_Sequence_Statistics_2/Mycobacterium_leprae.fasta")
is.vector(read_leprae)
# need to get it into the right format (just nucleotides)
lepraeseq <- read_leprae[[1]]
slidingwindowplot(200, lepraeseq)
slidingwindowplot(20000, lepraeseq)
slidingwindowplot(200000, lepraeseq)
# it's got very low GC at the beginning, 1,000,000 and 25M
```


#Q3. Write a function to calculate the AT content of a DNA sequence (ie. the fraction of the nucleotides in the sequence that are As or Ts). What is the AT content of the Mycobacterium leprae TN genome?
```{r}
#length(lepraeseq)
#leprae_table <- count(lepraeseq,1)
#leprae_table
#leprae_table[1]
#leprae_table[4]
AT_calculator <- function(genome)
{
  genome_table <- count(genome,1)
  A_count <- genome_table[1]
  T_count <- genome_table[4]
  100 * (A_count + T_count)/length(lepraeseq)
}
AT_calculator(lepraeseq)
```

#Q4. Write a function to draw a sliding window plot of AT content. Use it to make a sliding window plot of AT content along the Mycobacterium leprae TN genome, using a windowsize of 20000 nucleotides. Do you notice any relationship between the sliding window plot of GC content along the Mycobacterium leprae genome, and the sliding window plot of AT content?
```{r}
ATslidingwindow <- function(windowsize, genome)
{
  starts <- seq(1, length(genome)-windowsize, by = windowsize)
  n <- length(starts)
  chunkATs <- numeric(n)
  for (i in 1:n)
  {
    chunk <- genome[starts[i]:(starts[i]+windowsize-1)]
    chunkAT <- AT_calculator(chunk)
    print(chunkAT)
    chunkATs[i] <- chunkAT
  }
  plot(starts,chunkATs,type="b",xlab="Nucleotide start position",ylab="AT content")
}
ATslidingwindow(20000, lepraeseq)
```

#Q5. Is the 3-nucleotide word GAC over-represented or under-represented in the Mycobacterium leprae TN genome sequence?
```{r}
count(lepraeseq, 1)
# get G frequency
950202/(687041+938713+950202+692247)
# get A frequency
687041/(687041+938713+950202+692247)
# get C frequency
938713/(687041+938713+950202+692247)
count(lepraeseq, 3)
# get trinucleotide sum
# yuck, it looks like we're supposed to sum that all by hand ... NO WAY
trinucleotides <- count(lepraeseq, 3)
n <- length(trinucleotides)
trinucleotide_sum <- 0
for (i in 1:n)
{
  trinucleotide_sum <- trinucleotide_sum + trinucleotides[[i]]                
}
trinucleotide_sum
# get GAC frequency
GAC_frequency <- 61174/trinucleotide_sum
# GAC is slightly over-represented
GAC_frequency/(0.2907414 * 0.2102198 * 0.287226)
```

















