---
title: "DNA_Sequence_Statistics_1"
author: "Dave_Halvorsen"
date: "August 7, 2018"
output: html_document
---
This is my attempt at getting the provided code to work.
```{r}
# I am working on the 'Retrieving genome sequence data using SeqinR' section from 
# page 15 (PDF page 19) of 'A Little Book of R for Bioinformatics, Release 0.1'.

# This is the error message I get when I run the code (below)
# Error in socketConnection(host = host, port = port, server = server, blocking = blocking,  : 
# cannot open the connection
# In addition: Warning message:
# In socketConnection(host = host, port = port, server = server, blocking = blocking,  :
# pbil.univ-lyon1.fr:5558 cannot be opened
# Error in choosebank(db, timeout = 120) : 
# I wasn't able to open the socket connection:
#   o Check that your are connected to the internet.
#   o Check that port 5558 is not closed by a firewall.
#   o Try to increase timeout value (current is 5 seconds).
# My OS is Ubuntu 16.04. I tried updating the original timeout=5 to timeout=120.
# I tried 'sudo ufw allow 5558' and no change.  
# tried 'sudo iptables -A INPUT -m state --state NEW -m tcp -p tcp --dport 5558 -j ACCEPT
```


```{r}
getncbiseq <- function(accession)
{
  require("seqinr") # this function requires the SeqinR R package
  # first find which ACNUC database the accession is stored in:
  dbs <- c("genbank","regseq","refseqViruses","bacterial")
  numdbs <- length(dbs)
  for (i in 1:numdbs)
  {
    db <- dbs[i]
    choosebank(db, timeout=60, host = "pbil.univ-lyon1.fr", port = 5558)
    # check if the sequence is in ACNUC database 'db':
    resquery <- try(query(".tmpquery", paste("AC=", accession)), silent = TRUE)
    if (!(inherits(resquery, "try-error")))
    {
      queryname <- "query2"
      thequery <- paste("AC=",accession,sep="")
      query('queryname','thequery')
      # see if a sequence was retrieved:
      seq <- getSequence(query2$req[[1]])
      closebank()
      return(seq)
    }
    closebank()
  }
  print(paste("ERROR: accession",accession,"was not found"))
}
```

```{r}
dengueseq <- getncbiseq("NC_001477")
```
8/7/2017 ^ That code wasn't working, so I've skipped that section and just grabbed the sequence from NCBI. SKIP AHEAD to "Here's reading in the fasta file:". See email from Jean Lobry for more info ... they restarted teh server cause kit had problmes
```{r}
where.is.this.acc("NC_001477", stopAtFirst = FALSE)
```
CHANGING CODE! STARTING FOR LOOP AT 3 BECAUSE "refseqViruses" is where it's stored! :D
```{r}
getncbiseq <- function(accession)
{
  require("seqinr") # this function requires the SeqinR R package
  # first find which ACNUC database the accession is stored in:
  dbs <- c("genbank","regseq","refseqViruses","bacterial")
  numdbs <- length(dbs)
  for (i in 3:numdbs)
  {
    db <- dbs[i]
    choosebank(db, timeout=60, host = "pbil.univ-lyon1.fr", port = 5558)
    # check if the sequence is in ACNUC database 'db':
    resquery <- try(query(".tmpquery", paste("AC=", accession)), silent = TRUE)
    if (!(inherits(resquery, "try-error")))
    {
      queryname <- "query2"
      thequery <- paste("AC=",accession,sep="")
      query('queryname','thequery')
      # see if a sequence was retrieved:
      seq <- getSequence(query2$req[[1]])
      closebank()
      return(seq)
    }
    closebank()
  }
  print(paste("ERROR: accession",accession,"was not found"))
  
}
dengueseq <- getncbiseq("NC_001477")
```
refactored code suggestion from Stack Overflow: https://stackoverflow.com/questions/37856007/error-with-a-function-to-retrieve-data-from-a-database
Also, I changed it to start at 3 (see above)
```{r}
getncbiseq <- function(accession)
{
require("seqinr") # this function requires the SeqinR R package
# first find which ACNUC database the accession is stored in:
dbs <- c("genbank","refseq","refseqViruses","bacterial")
numdbs <- length(dbs)
for (i in 3:numdbs)
{
db <- dbs[i]
choosebank(db)
# check if the sequence is in ACNUC database 'db':
resquery <- try(query(".tmpquery", paste("AC=", accession)), silent = TRUE)

if (!(inherits(resquery, "try-error"))) {
  queryname <- "query2"
  thequery <- paste("AC=", accession, sep="")
  query2 <- query(queryname, thequery)
  # see if a sequence was retrieved:
  seq <- getSequence(query2$req[[1]])
  closebank()
  return(seq)
}
closebank()
}
print(paste("ERROR: accession",accession,"was not found"))
}
getncbi_dengueseq <- getncbiseq("NC_001477")
```

```{r}
# still of type character
typeof(getncbi_dengueseq)
getncbi_dengueseq
```
Awesome! It works now :)
```{r}
getncbi_dengueseq[1:50]
```
writing out a fasta file
```{r}
write.fasta(names="ncbiDEN-1", sequences=getncbi_dengueseq, file.out="ncbi_den1.fasta")
```
reading in a fasta file (the file from above)
```{r}
read_dengue <- read.fasta(file="/media/david/Linux/R_for_Bioinformatics/ncbi_den1.fasta")
```

```{r}
# this is the WHOLE FILE
read_dengue
dengue_sequence <- read_dengue[[1]]
# this is just the sequence!
dengue_sequence
```

```{r}
# this is the length of the dengue sequence
length(dengue_sequence)
# this is the # of different nucleotides
table(dengue_sequence)
# what is the GC content?
100*(2240+2770)/(2240+2770+3426+2299)
# alternatively, there's a function to do that:
GC(dengue_sequence)
```

```{r}
# here's how to get the frequencies of longer words (1 is the same as regular GC() function)
count(dengue_sequence, 1)
count(dengue_sequence, 2)
count(dengue_sequence, 3)
# The output of count is a table object, so you can use [[]] to grab singular values
count(dengue_sequence, 1)[[3]]
# if you store the cout function as a variable, you can grab items by name:
dengue_count_1 <- count(dengue_sequence, 1)
dengue_count_1[["g"]]
```

exercises from the first chapter
```{r}
# Q1. What are the last twenty nucleotides of the Dengue virus genome sequence?

# Q2. What is the length in nucleotides of the genome sequence for the bacterium Mycobacterium leprae strain TN?

# Q3. How many of each of the four nucleotides A, C, T, and G, and any other symbols, are there in the Mycobacterium leprae

# Q4. What is the GC content of the Mycobacterium leprae TN genome sequence, when i
```

Here's more stuff on seqinr that I planned to do if I couldn't get the main exercises to work ... no need to do it now! However, I did find some others who improvised some exercises:
https://github.com/haruosuz/r4bioinfo/tree/master/R_seqinR#exercise-1
https://github.com/haruosuz/r4bioinfo/tree/master/R_Avril_Coghlan#dna-sequence-statistics-1
https://rstudio-pubs-static.s3.amazonaws.com/73927_60e9fb345d9d4175b8994b1ff1eff2e2.html



