---
title: "Sequence_Databases"
author: "Dave_Halvorsen"
date: "August 12, 2018"
output:
  pdf_document: default
  html_document: default
---

```{r}
# this'll show a list of all the databases I can access with seqinr
library("seqinr")
# the choosebank() code was causing a problem downstream, so it's commented out
# choosebank()
```

```{r}
# the book calls for all of this to be run, BUT I'm concerned it'll
# interfere with further steps, so I've commented it out
# specify we want to search the 'genbank' ACNUC sub-database
# choosebank("genbank")
# specify a search for 'refseq'
# choosebank('refseq') 
# queries need a name and type
# query("RefSeqBact", "SP=Bacteria")
# you need to finally close the database
# closebank()
# 3 step review: use "choosebank()" to select sub-database, 
# use "query()" to query, and 3rd use "closebank()"
```

```{r}
# search for DEN-1 virus genome
choosebank("refseqViruses")
# TYPO IN THE BOOK! it just list query onwards w/o the variable 
# and that doesn't work. This does:
Dengue1 <- query("Dengue1", "AC=NC_001477")
attributes(Dengue1)
# to get an objects attributes add $attribute to the object
Dengue1$nelem
# to get accession #
Dengue1$req
# get names, class
attr(Dengue1, "names")
attr(Dengue1, "class")
# this calls forth for the sequence data
dengueseq <- getSequence(Dengue1$req[[1]])
# first 50 elements
dengueseq[1:50]
# getting annotations
annots <- getAnnot(Dengue1$req[[1]])
# getting first 20 lines of annots
annots[1:20]
# close the database when you're done
closebank()
```

```{r}
# finding the sequences published in Nature 460:352-358
# DOES NOT WORK. Page 43 (PDF 47), so i've commented it out. Error is
# Error in query("naturepaper", "R=Nature/460/352") : invalid request:"unknown reference at (^): \"R
# Two examples of other's who've failed
# https://www.biostars.org/p/197312/#332392
# http://lists.r-forge.r-project.org/pipermail/seqinr-forum/2017q3/000252.html
# ^above forum suggests the Nature paper in question might not exist ... 
# specifying we want genbank
# choosebank("genbank")
# search criteria for Nature
# query('naturepaper', 'R=Nature/460/352')
# naturepaper$nelem
```

```{r}
# trying my version ... 
# couldn't get it to work, so I've commented it out
# even the query help page example of 'JMB/13/5432' doesn't work!
# naturepaper <- query('naturepaper', 'R=JMB/13/5432')
# I get this error message with this normal seeming request
# Error in query("naturepaper", "j=Nature", "Y=2006") : argument socket = Y=2006 
# is not a socket connection.
# naturepaper <- query('naturepaper', 'j=Nature', 'Y=2006')
# naturepaper$nelem

# I've repliacted the functional Dengue1 code from above
choosebank("refseqViruses")
dengue_attempt <- query("Dengue1", 'AC=NC_001477')
closebank()
```

```{r}
# the suggested code from page 44 (PDF 48) doesn't work either. Error:
# Error in getSequence(humtRNAs) : object 'humtRNAs' not found
# I'm growing tired of all this dysfunctional code. The Rosalind.info
# site may be a better usage of my time than fumbling through this mess
# choosebank("genbank")
# query("humtRNAs", "SP=homo sapiens AND M=TRNA")
# myseqs <- getSequence(humtRNAs)
# mynames <- getName(humtRNAs)
# write.fasta(myseqs, mynames, file.out="humantRNAs.fasta")
# closebank()
```
# Q1 What information about the rabies virus sequence (NCBI accession NC_001542) can you obtain from its annotations in the NCBI Sequence Database?
```{r}
# go to http://www.ncbi.nlm.nih.gov/
# use accession 'NC_001542' to find the rabies virus. 
# since it's a virus, it'll be in 'refseqViruses'. Use the query format to grab it.
# load the required package
library("seqinr")   
# select the virus database
choosebank("refseqViruses")
# specify the query
rabies <- query("rabies", "AC=NC_001542")
# retrieve the annotations
annots <- getAnnot(rabies$req[[1]])
annots[1:20]
closebank()
```
# Q2 How many nucleotide sequences are there from the bacterium Chlamydia trachomatis in the NCBI Sequence Database?
```{r}
library("seqinr")
# selecting genbank for the searching
choosebank("genbank")
Ctrachomatis <- query("Ctrachomatis", "SP=Chlamydia trachomatis")
Ctrachomatis$nelem
closebank()
```
#Q3 How many nucleotide sequences are there from the bacterium Chlamydia trachomatis in the RefSeq part of the NCBI Sequence Database?
```{r}
library("seqinr")
# looking in refseq
choosebank("refseq")
Ctrachomatis2 <- query("Ctrachomatis2", "SP=Chlamydia trachomatis")
Ctrachomatis2$nelem
closebank()
# looking in bacterial
# NVM, it's off for maintenance ... commenting out
# choosebank("bacterial")
# Ctrachomatis2 <- query("Ctrachomatis2", "SP=Chlamydia trachomatis")
# Ctrachomatis2$nelem
# closebank()
```
# Q4 How many nucleotide sequences were submitted to NCBI by Matthew Berriman?
```{r}
library("seqinr")
choosebank("genbank")
mberriman <- query("mberriman", "AU=Berriman")
mberriman$nelem
closebank()
```
# Q5 How many nucleotide sequences from the nematode worms are there in the RefSeq Database?
```{r}
library("seqinr")
choosebank("refseq")
nematodes <- query("nematodes", "SP=Nematoda")
nematodes$nelem
closebank()
```
# Q6 How many nucleotide sequences for collagen genes from nematode worms are there in the NCBI Database?
```{r}
library("seqinr")
choosebank("genbank")
collagen <- query("collagen", "SP=Nematoda AND K=collagen")
collagen$nelem
closebank()
```
# Q7 How many mRNA sequences for collagen genes from nematode worms are there in the NCBI Database?
```{r}
library("seqinr")
choosebank("genbank")
collagen2 <- query("collagen2", "SP=Nematoda AND K=collagen AND M=mRNA")
collagen2$nelem
closebank()
```
# Q8 How many protein sequences for collagen proteins from nematode worms are there in the NCBI database?
```{r}
# need to search ncbi.nlm.nih.gov website for "Nematoda[ORGN] AND collagen"
# result: 1 to 20 of 5563 Found 12361 nucleotide sequences. Nucleotide (5563) EST (6798)
# there isn't an ACNUC database for this
```
# Q9 What is the accession number for the Trypanosoma cruzi genome in NCBI?
```{r}
# go to ncbi.nlm.nih.gov and search '"Trypanosoma cruzi"[ORGN]'
# accession is NZ_AAHK00000000.1
```
# Q10 How many fully sequenced nematode worm species are represented in the NCBI Genome database?
```{r}
# ncbi Genome search term 'Nematoda[ORGN]'
# 1 to 20 of 102
```

# REVISION EXERCISES 1
# Q1. What is the length of (total number of base-pairs in) the Schistosoma mansoni mitochondrial genome (NCBI accession NC_002545), and how many As, Cs, Gs and Ts does it contain?
```{r}
# search ncbi.nlm.nih.gov for "NC_002545"
# send to file FASTA
# read.fasta requires the seqinr library
library("seqinr")
smansoni <- read.fasta(file="smansoni.fasta")
# smansoni <- read.fasta(file="/media/david/Linux/R_for_Bioinformatics/Sequence_Databases/smansoni.fasta")
# limit the data to only the sequence
smansoniseq <- smansoni[[1]]
# measure the sequence length
length(smansoniseq)
# count the nucleotides
table(smansoniseq)
```
# Q2. What is the length of the Brugia malayi mitochondrial genome (NCBI accession NC_004298), and how many As, Cs, Gs and Ts does it contain?
```{r}
bmalayi <- read.fasta(file="bmalayi.fasta")
bmalayiseq <- bmalayi[[1]]
length(bmalayiseq)
table(bmalayiseq)
```
# Q3. What is the probability of the Brugia malayi mitochondrial genome sequence (NCBI accession NC_004298), according to a multinomial model in which the probabilities of As, Cs, Gs and Ts (pA, pC, pG, and pT) are set equal to the fraction of As, Cs, Gs and Ts in the Schistosoma mansoni mitochondrial genome?
```{r}
# getting the counts of smansoniseq
mytable <- table(smansoniseq)
mytable
mytable <- mytable/length(smansoniseq)
# get the frequencies for different letters for smansoniseq
freqA <- mytable[["a"]]
freqC <- mytable[["c"]]
freqG <- mytable[["g"]]
freqT <- mytable[["t"]]
# create a vector of the frequencies of a, c, g, t
probabilities <- c(freqA, freqC, freqG, freqT)
probabilities
```

```{r}
# this function calculates the probability of the sequence being in the order that it is
# obviously, the probability will be infinitesimally small ...
multinomialprob <- function(mysequence, probabilities)
{
  # these are the options for sequence letter
  nucleotides <- c("A", "C", "G", "T")
  # adding nucleotide letter labels for the probabilities
  names(probabilities) <- nucleotides
  # making the sequence uppercase
  mysequence <- toupper(mysequence)
  # get the length of the sequence
  seqlength <- length(mysequence)
  # make an empty variable to hold the probability of the sequence (as it's calculated)
  seqprob <- numeric()
  # cycle through each nucleotide letter of the sequence
  for (i in 1:seqlength)
  {
    # store the current nucleotide
    nucleotide <- mysequence[i]
    # store the probability for the current nucleotide
    nucleotideprob <- probabilities[nucleotide]
    # for the FIRST nucleotide there isn't an ongoing proability to multiply by
    # so it's just the probability of the current nucleotide
    if (i == 1) { seqprob <- nucleotideprob[[1]] }
    # for all other positions, the current nucleotide probability is multiplied by the entire sequence
    else { seqprob <- seqprob * nucleotideprob[[1]] }
  }
  return(seqprob)
}
# call the function on the bmalyaiseq sequence
multinomialprob(bmalayiseq, probabilities)
```

# Q4. What are the top three most frequent 4-bp words (4-mers) in the genome of the bacterium Chlamydia trachomatis strain D/UW-3/CX (NCBI accession NC_000117), and how many times do they occur in its sequence?
```{r}
# initially, I downloaded the file from ncbi.nlm.gov
ctrachomatis <- read.fasta(file="ctrachomatis.fasta")
ctrachomatis
# but I'd like more practice w/ R, so here's using query to get it
# NEVERMIND, the bacterial database is currently offline for maintenance
# Error message:
# Error in acnucopen(bank, socket) : Database with name -->bacterial<-- is currently off for maintenance, please try again later.
# commented it out
library("seqinr")
# choosebank("bacterial")
# query("ctrachomatis", "AC=NC_000117")
# getting the sequence
# this is just if you query it (diff format)
# ctrachomatisseq <- getSequence(ctrachomatis$req[[1]])
# closebank()
# getting the sequence
ctrachomatisseq <- ctrachomatis[[1]]
# getting 4 counts
mytable <- count(ctrachomatisseq, 4)
# sorting the counts (last is highest)
sorted_mytable <- sort(mytable)
sorted_mytable
# length of sorted table
length_sorted_mytable <- length(sorted_mytable)
length_sorted_mytable
# slicing the list with the actual numbers OR using the length and length-3
sorted_mytable[(length_sorted_mytable-2):length_sorted_mytable]
sorted_mytable[254:256]
typeof(ctrachomatisseq)
```
# Q5. Write an R function to generate a random DNA sequence that is n letters long (that is, n bases long) using a multinomial model in which the probabilities pA, pC, pG, and pT are set equal to the fraction of As, Cs, Gs and Ts in the Schistosoma mansoni mitochondrial genome.
```{r}
table_smanson <- table(smansoniseq)
table_smanson
length_smanson <- length(smansoniseq)
length_smanson
smanson_frequencies <- table_smanson/length_smanson
A_freq <- smanson_frequencies[["a"]]
C_freq <- smanson_frequencies[["c"]]
G_freq <- smanson_frequencies[["g"]]
T_freq <- smanson_frequencies[["t"]]
smanson_probabilities <- c(A_freq, C_freq, G_freq, T_freq)
smanson_probabilities
nucleotides <- c("A", "C", "G", "T")
# adding nucleotide letter labels for the probabilities
names(smanson_probabilities) <- nucleotides
# smanson_probabilities

random_DNA_gen <- function(n) 
{
# calculating probabilities of smanson nucleotides
# smanson_probabilities
A <- smanson_probabilities[["A"]]
C <- smanson_probabilities[["C"]]
G <- smanson_probabilities[["G"]]
T <- smanson_probabilities[["T"]]
# checking the values
# A
# C
# G
# T
# they sum to 1
# sum(A, C, G, T)
nucleotide_vector = character()
n <- 5
for (i in 1:n)
{
  print("random number: ")
  random_number <- runif(1, 0, 1)
  print(random_number)
  print("random nucleotide: ")
  if (random_number <= 0.253486) 
    {
      random_nucleotide <- "A"
    }
  else if (random_number <= (0.253486+0.08518904)) 
    {
      random_nucleotide <- "C"
    }
  else if (random_number <= (0.253486+0.08518904+0.2294138)) 
    {
      random_nucleotide <- "G"
    }
  else 
    {
      random_nucleotide <- "T"
    }
  # print(random_nucleotide)
  nucleotide_vector[i] = random_nucleotide
  }
  print(nucleotide_vector)
}

random_DNA_gen(5)
```
# Q6. Give an example of using your function from Q5 to calculate a random sequence that is 20 letters long, using a multinomial model with pA =0.28, pC =0.21, pG =0.22, and pT =0.29.
```{r}
# updated function from before to take an input of nucleotide frequencies
random_DNA_PROB_gen <- function(n, pA, pC, pG, pT) 
{
  # check to make sure that the sum of probabilities == 1
  stopifnot(sum(pA + pC + pG + pT)==1)
  # create an empty character vector for the nucleotide string
  nucleotide_vector = character()
  # run the for loop from 1 until n
  for (i in 1:n)
  {
    # print("random number: ")
    # generate a random number between 0 and 1
    random_number <- runif(1, 0, 1)
    # print(random_number)
    # print("random nucleotide: ")
    # probability tree of increasing sums of the nucleotide frequency
    # this works because it starts at the lowest numbers and goes up 
    # it wouldn't work in the opposite direction (low to high) cause everything 
    # would sort into the lowest
    if (random_number <= pA) 
      {
        random_nucleotide <- "A"
      }
    else if (random_number <= (pA + pC)) 
      {
        random_nucleotide <- "C"
      }
    else if (random_number <= (pA + pC + pG)) 
      {
        random_nucleotide <- "G"
    }
    # the probability sum isn't necessary here because higher is the only option
    # this function doesn't take into account errors in user input ... future might
    # wanna change to an else going to error and having G be assigned in an else if
    else 
      {
        random_nucleotide <- "T"
      }
  # print(random_nucleotide)
  nucleotide_vector[i] = random_nucleotide
  }
  print(nucleotide_vector)
}

random_DNA_PROB_gen(20, 0.28, 0.21, 0.22, 0.29)
```
# Q7. How many protein sequences from rabies virus are there in the NCBI Protein database?
```{r}
# ncbi.nlm.nih.gov
# search protein database for 'rabies virus'
# 30,211 was the # returned
```
# Q8. What is the NCBI accession for the Mokola virus genome?
```{r}
# ncbi.nlm.nih.gov
# search genome database for 'Mokola virus'
# NC_006429.1
```










