---
title: "Multiple Alignment and Phylogenetic Trees"
output: github_document
---

```{r}
# this function can download multiple sequences from a specified database
# the code provided by the book is defunct
# see StackOverflow for the proper code: 
# https://stackoverflow.com/questions/37856007/error-with-a-function-to-retrieve-data-from-a-database 
retrieveseqs <- function(seqnames,acnucdb)
{
  # this is a list that the sequences will be stored in
  myseqs <- list()
  require("seqinr")
  choosebank(acnucdb)
  for (i in 1:length(seqnames))
  {
    seqname <- seqnames[i]
    print(paste("Retrieving sequence",seqname,"..."))
    queryname <- "query2"
    thequery <- paste("AC=",seqname,sep="")
    # I have changed this function every so slightly from it's old state (above stackoverflow)
    # page 77 (PDF page 81) calls this function with UniProt accession E1FUV2
    # HOWEVER, E1FUV2 IS NOT on UniProt anymore, so this function fails when it can't
    # find it. I have saved the initial state of the program here:
    # query2 <- query(queryname, thequery)
    # this changed version calls query within a tryCatch statement
    # without the try it causes this error message: 
    # Error in query(queryname, thequery) : invalid request:
    # "unknown accession number at (^): \"AC
    query2 <- try(query(queryname, thequery))
    # create vector "seq" containing the sequence
    # also, I added a try block to this initial code ... do I really need the above try?
    # YES, I need the above try! I copied the error message that it causes up there^
    # seq <- getSequence(query2$req[[1]])
    # if I leave out the following try, I get this error message:
    # Error: $ operator is invalid for atomic vectors
    seq <- try(getSequence(query2$req[[1]]))
    myseqs[[i]] <- seq
  }
  closebank()
  return(myseqs)
}
```

```{r}
# retrieving protein sequences, from UniProt, for rabies virus phosphoprotein, Mokola virus 
# phosphoprotein, Lagos bat virus phosphoprotein, and Western Caucasion bat virus phosphoprotein
# this is the list of protein sequences to download
seqnames <- c("P06747", "P0C569", "O56773", "Q5VKP1")
# saving the protein sequences into a vector
seqs <- retrieveseqs(seqnames,"swissprot")
# it is 4 long
length(seqs)
# saving JUST the sequence part
seq1 <- seqs[[1]]
# printing out the first 20 protein letters
seq1[1:20]
# saving JUST the sequence part
seq2 <-seqs[[2]]
# printing out the first 20 protein letters
seq1[1:20]
```

```{r}
# saving the list of protein reads into fasta format
write.fasta(seqs, seqnames, file="phosphoproteins.fasta")
```

```{r}
# completed a sequence alignment with the clustal program (see pg 66 of little book of r 
# for bioinformatics)
# reading in the aligned sequences
virusaln <- read.alignment(file = "phosphoproteins.phy", format = "phylip")
# the alignments are very long, so only part of them is shown
virusaln$seq
```

```{r}
# This function will print out the alignment over multiple lines
printMultipleAlignment <- function(alignment, chunksize=60)
{
  # this function requires the Biostrings package
  require("Biostrings")
  # find the number of sequences in the alignment
  numseqs <- alignment$nb
  # find the length of the alignment
  alignmentlen <- nchar(alignment$seq[[1]])
  starts <- seq(1, alignmentlen, by=chunksize)
  n <- length(starts)
  # get the alignment for each of the sequences:
  aln <- vector()
  lettersprinted <- vector()
  for (j in 1:numseqs)
  {
    alignmentj <- alignment$seq[[j]]
    aln[j] <- alignmentj
    lettersprinted[j] <- 0
  }
  # print out the alignment in blocks of 'chunksize' columns:
  for (i in 1:n) 
  {
    for (j in 1:numseqs)
    {
      alnj <- aln[j]
      chunkseqjaln <- substring(alnj, starts[i], starts[i]+chunksize-1)
      chunkseqjaln <- toupper(chunkseqjaln)
      # Find out how many gaps there are in chunkseqjaln:
      # countPattern() is from Biostrings package
      gapsj <- countPattern("-",chunkseqjaln) 
      # Calculate how many residues of the first sequence we have
      # printed so far in the alignment:
      lettersprinted[j] <- lettersprinted[j] + chunksize - gapsj
      print(paste(chunkseqjaln,lettersprinted[j]))
    }
    print(paste(' '))
  }
}
```

```{r}
# print out the multiple alignment of virus phosphoproteins in 60 column blocks
printMultipleAlignment(virusaln, 60)
```

```{r}
# it's important to discard areas that are not homologous
cleanAlignment <- function(alignment, minpcnongap, minpcid)
{
  # make a copy of the alignment to store the new alignment in:
  newalignment <- alignment
  # find the number of sequences in the alignment
  numseqs <- alignment$nb
  # empty the alignment in "newalignment")
  for (j in 1:numseqs) { newalignment$seq[[j]] <- "" }
  # find the length of the alignment
  alignmentlen <- nchar(alignment$seq[[1]])
  # look at each column of the alignment in turn:
  for (i in 1:alignmentlen)
  {
    # see what percent of the letters in this column are non-gaps:
    nongap <- 0
    for (j in 1:numseqs)
    {
      seqj <- alignment$seq[[j]]
      letterij <- substr(seqj,i,i)
      if (letterij != "-") { nongap <- nongap + 1}
    }
    pcnongap <- (nongap*100)/numseqs
    # Only consider this column if at least minpcnongap % of the letters are not gaps:
    if (pcnongap >= minpcnongap)
      {
        # see what percent of the pairs of letters in this column are identical:
        numpairs <- 0; numid <- 0
        # find the letters in all of the sequences in this column:
        for (j in 1:(numseqs-1))
        {
          seqj <- alignment$seq[[j]]
          letterij <- substr(seqj,i,i)
          for (k in (j+1):numseqs)
          {
              seqk <- alignment$seq[[k]]
              letterkj <- substr(seqk,i,i)
              if (letterij != "-" && letterkj != "-")
            {
              numpairs <- numpairs + 1
              if (letterij == letterkj) 
            { 
              numid <- numid + 1}
            }
          }
      }
      pcid <- (numid*100)/(numpairs)
      # Only consider this column if at least %minpcid of the pairs of letters are identical:
      if (pcid >= minpcid)
        {
        for (j in 1:numseqs)
        {
          seqj <- alignment$seq[[j]]
          letterij <- substr(seqj,i,i)
          newalignmentj <- newalignment$seq[[j]]
          newalignmentj <- paste(newalignmentj,letterij,sep="")
          newalignment$seq[[j]] <- newalignmentj
        }
      }
    }
  }
  return(newalignment)
}
```

```{r}
# printing out the well conserved areas
cleanedvirusaln <- cleanAlignment(virusaln, 30, 30)
printMultipleAlignment(cleanedvirusaln)
```

```{r}
# calculate the distance between the alignments
virusdist <- dist.alignment(virusaln)
virusdist
```

```{r}
# make a vector containing the names of the sequences
seqnames <- c("AF049118", "AF049114", "AF049119", "AF049115")
# Retrieve the sequences and store them in list variable "seqs"
seqs <- retrieveseqs(seqnames,"genbank")
# save the sequences into a fasta
write.fasta(seqs, seqnames, file="virusmRNA.fasta")
```

```{r}
# read the alignment into R
virusmRNAaln <- read.alignment(file = "virusmRNA.phy", format = "phylip")
# calculate a genetic distance matrix based on a protein sequence alignment

# require the 'ape' package
library('ape')
# convert the alignment to DNAbin
virusmRNAalnbin <- as.DNAbin(virusmRNAaln)
# calculate the genetic distance matrix
virusmRNAdist <- dist.dna(virusmRNAalnbin)
# Print out the genetic distance matrix
virusmRNAdist
```

```{r}
# the unrootedNJtree() function calculates pairwise distances betwen the sequences
# based on the alignment
# it builds a phylogenetic tree based on those distances
unrootedNJtree <- function(alignment,type)
{
  # this function requires the ape and seqinR packages:
  require("ape")
  require("seqinr")
  # define a function for making a tree:
  makemytree <- function(alignmentmat)
  {
    alignment <- ape::as.alignment(alignmentmat)
    if (type == "protein")
    {
      mydist <- dist.alignment(alignment)
    }
    else if (type == "DNA")
    {
      alignmentbin <- as.DNAbin(alignment)
      mydist <- dist.dna(alignmentbin)
    }
    mytree <- nj(mydist)
    # get rid of spaces in tip names
    mytree <- makeLabel(mytree, space="")
    return(mytree)
  }
  # infer a tree
  mymat <- as.matrix.alignment(alignment)
  mytree <- makemytree(mymat)
  # bootstrap the tree
  myboot <- boot.phylo(mytree, mymat, makemytree)
  # plot the tree:
  # plot the unrooted phylogenetic tree
  plot.phylo(mytree,type="u")
  # plot the bootstrap values
  nodelabels(myboot,cex=0.7)
  # make the bootstrap values be the node labels
  mytree$node.label <- myboot
  return(mytree)
}
```

```{r}
# make a tree
virusalntree <- unrootedNJtree(virusaln,type="protein")
```

```{r}
# make a filtered copy of the multple alignment
cleanedvirusalntree <- unrootedNJtree(cleanedvirusaln,type="protein")
```

```{r}
# retrieving sequences from UniProt
# this was the initial seqnames:
# seqnames <- c("Q10572","E3M2K8","Q8WS01","E1FUV2","A8NSK3","Q9VT99")
# HOWEVER,
# E1FUV2 was removed from UniProt on October 3, 2012. https://www.uniprot.org/uniprot/E1FUV2
# I have removed E1FUV2 from seqnames here ... I downloaded the fasta file manually
# ... I might add it back into the mix. Alternatively, I could use a try statement
# in the retrieveseqs function. Yup, ^the above was gonna be a lazy option. I figured it out
# see up top at line 11 for my fix; I used try() around the bad accessions and the function ran
# HOWEVER, now I have two empty spots in my list of seqs ... actually, they're not empty
# they're filled with the error code.
# I also got A8NSK3. It was deleted on September 16, 2015 this entry was deleted.
seqnames <- c("Q10572","E3M2K8","Q8WS01","E1FUV2","A8NSK3","Q9VT99")
seqs <- retrieveseqs(seqnames,"swissprot")
write.fasta(seqs, seqnames, file="fox1.fasta")
```

```{r}
# the list of names is 6 long, as expected
length(seqs)
# the right names are there as expected
seqnames
# it's a list
typeof(seqs)
# 4 and 5 (E1FUV2 and A8NSK3) caused the accession # error
# seqs[[1]]
# seqs[[2]]
# seqs[[3]]
# E1FUV2 ... I need to manually get this
seqs[[4]]
# A8NSK3 ... I need to manually get this
seqs[[5]] 
# seqs[[6]]
# the I've saved the old fasta file of fox1.fasta as a new name
# fox1_old_ERROR.fasta.
# I have manually added the proper fasta proteins to fox1.fasta7
```

```{r}
# see above code for the issues with E1FUV2 and A8NSK3
# I have since manually added the missing sequences to the fox1.fasta file
# I ran this fasta through Clustal
fox1aln <- read.alignment(file = "fox1.phy", format = "phylip")
```

```{r}
# building a rooted tree out of proteins
rootedNJtree <- function(alignment, theoutgroup, type)
{
# load the ape and seqinR packages:
require("ape")
require("seqinr")
# define a function for making a tree:
makemytree <- function(alignmentmat, outgroup=`theoutgroup`)
  {
    alignment <- ape::as.alignment(alignmentmat)
    if (type == "protein")
    {
      mydist <- dist.alignment(alignment)
    }
    else if (type == "DNA")
    {
      alignmentbin <- as.DNAbin(alignment)
      mydist <- dist.dna(alignmentbin)
    }
    mytree <- nj(mydist)
    mytree <- makeLabel(mytree, space="") # get rid of spaces in tip names.
    myrootedtree <- root(mytree, outgroup, r=TRUE)
    return(myrootedtree)
  }
  # infer a tree
  mymat <- as.matrix.alignment(alignment)
  myrootedtree <- makemytree(mymat, outgroup=theoutgroup)
  # bootstrap the tree
  myboot <- boot.phylo(myrootedtree, mymat, makemytree)
  # plot the tree:
  plot.phylo(myrootedtree, type="p") # plot the rooted phylogenetic tree
  nodelabels(myboot,cex=0.7)
  # plot the bootstrap values
  mytree$node.label <- myboot
  # make the bootstrap values be the node labels
  return(mytree)
}
```


```{r}
fox1alntree <- rootedNJtree(fox1aln, "Q9VT99",type="protein")
```

```{r}
virusmRNAaln <- read.alignment(file = "virusmRNA.phy", format = "phylip")
virusmRNAalntree <- unrootedNJtree(virusmRNAaln, type="DNA")
```

```{r}
# phylogenetic trees are typically stored in the 'Newick-format'
write.tree(virusmRNAalntree, "virusmRNA.tre")
```

# Q1. Calculate the genetic distances between the following NS1 proteins from different Dengue virus strains: Dengue virus 1 NS1 protein (Uniprot Q9YRR4), Dengue virus 2 NS1 protein (UniProt Q9YP96), Dengue virus 3 NS1 protein (UniProt B0LSS3), and Dengue virus 4 NS1 protein (UniProt Q6TFL5). Which are the most closely related proteins, and which are the least closely related, based on the genetic distances?
```{r}
# calculate the genetic distances between the following NS1 proteins
# Dengue virus 1 NS1 protein (Uniprot Q9YRR4)
# Dengue virus 2 NS1 protein (UniProt Q9YP96)
# Dengue virus 3 NS1 protein (UniProt B0LSS3)
# Dengue virus 4 NS1 protein (UniProt Q6TFL5)

# code to download the protein sequences
viralseqnames <- c("Q9YRR4","Q9YP96","B0LSS3","Q6TFL5")
viralseqs <- retrieveseqs(viralseqnames,"swissprot")
write.fasta(viralseqs, viralseqnames, file="dengue_strains.fasta")
```
```{r}
# use clustal to match the protein and save as "dengue_strains.phy"
# I think this code will work ... I changed the variable names
# read the alignment into R
viralmRNAaln <- read.alignment(file = "dengue_strains.phy", format = "phylip")
# calculate a genetic distance matrix based on a protein sequence alignment
# calculate the genetic distance matrix
viralmRNAdist <- dist.alignment(viralmRNAaln)
# Print out the genetic distance matrix
viralmRNAdist
```

# Q2. Build an unrooted phylogenetic tree of the NS1 proteins from Dengue virus 1, Dengue virus 2, Dengue virus 3 and Dengue virus 4, using the neighbour-joining algorithm. Which are the most closely related proteins, based on the tree? Based on the bootstrap values in the tree, how confident are you of this?
```{r}
viralmRNAtree <- unrootedNJtree(viralmRNAaln,type="protein")
```

# Q3. Build an unrooted phylogenetic tree of the NS1 proteins from Dengue viruses 1-4, based on a filtered alignment of the four proteins (keeping alignment columns in which at least 30% of letters are not gaps, and in which at least 30% of pairs of letters are identical). Does this differ from the tree based on the unfiltered alignment (in Q2)? Can you explain why?
```{r}
# clean up the sequence first
cleanedviralmRNAaln <- cleanAlignment(viralmRNAaln, 30, 30)
# build the tree w/ the cleaned up sequence
cleanedviralmRNAalntree <- unrootedNJtree(cleanedviralmRNAaln, type="protein")
```
```{r}
# lots of open gaps
printMultipleAlignment(viralmRNAaln)
```
```{r}
# the cleaned version has less gaps
printMultipleAlignment(cleanedviralmRNAaln)
```

# Q4. Q4. Build a rooted phylogenetic tree of the Dengue NS1 proteins based on a filtered alignment, using the Zika virus protein as the outgroup. Which are the most closely related Dengue virus proteins, based on the tree? What extra information does this tree tell you, compared to the unrooted tree in Q2?
```{r}
# grab sequences; all, but Q327E1 are old ... still getting them though
seqnames <- c("Q9YRR4", "Q9YP96", "B0LSS3", "Q6TFL5", "Q32ZE1")
# retrieve the sequences are store them
seqs <- retrieveseqs(seqnames, "swissprot")
# write the sequences out as a fasta file
write.fasta(seqs, seqnames, file="NS1b.fasta")
# outside of R, run this through Clustal
NS1baln <- read.alignment(file = "NS1b.phy", format = "phylip")
# get rid of the unreliable parts
cleanedNS1baln <- cleanAlignment(NS1baln, 30, 30)
# build a rooted phylogenetic tree w/ Zika virus (Q32ZE1) as outgroup
cleanedNS1balntree <- rootedNJtree(cleanedNS1baln, "Q32ZE1",type="protein")
#  Q9YP96 (Dengue virus 2 NS1) and Q6TFL5 (Dengue virus 4 NS1) are grouped together
```





